File: 2019/day4.py
from typing import Tuple

def match(nr: int) -> Tuple[bool, bool]:
    if nr < 10e4:
       return (False, False)

    streak: int = 1
    last_digit = None

    has_double_streak = False
    has_min2_streak = False

    for digit in str(nr):
        if last_digit is not None:
            if digit == last_digit:
               streak += 1
            else:
               if streak >= 2:
                  has_min2_streak = True
               if streak == 2:
                  has_double_streak = True
               streak = 1

            if int(digit) < int(last_digit):
               return (False, False)

        last_digit = digit

    if streak >= 2:
       has_min2_streak = True
    if streak == 2:
       has_double_streak = True

    return (has_min2_streak, has_double_streak)

# Part 1 tests
assert match(111111)[0]
assert not match(223450)[0]
assert not match(123789)[0]

# Part 2 tests
assert match(112233)[1]
assert not match(123444)[1]
assert match(111122)[1]

#
# Solve
#

amnt_part1: int = 0
amnt_part2: int = 0
for nr in range(137683, 596253):
    part1_match, part2_match = match(nr)
    if part1_match:
        amnt_part1 += 1
    if part2_match:
        amnt_part2 += 1

print(f"Part 1: {amnt_part1}")
print(f"Part 2: {amnt_part2}")

File: 2019/day4.c
#include <stdint.h>
#include <stdio.h>

#define FALSE 0u
#define STREAK_MIN2 1u
#define STREAK_EXACT2 2u

#define TEN 10
#define MIN_SIZE 100000u

#define INPUT_MIN 137683u
#define INPUT_MAX 596253u

int match(uint32_t nr)
{
    if (nr < MIN_SIZE) {
       return FALSE;
    }

    int streak = 1;
    int last_digit = -1;

    uint8_t return_value = 0;
    int digit;
    
    do {
        digit = nr % TEN;
        nr /= TEN;
       
        if (last_digit > -1) {
              if (digit == last_digit) {
                 streak += 1;
              } else {
                 if (streak >= 2) {
                    return_value |= STREAK_MIN2;
                 }
                 if (streak == 2) {
                    return_value |= STREAK_EXACT2;
                 }
                 streak = 1;
              }
              if (digit > last_digit) {
                 return FALSE;
              }
        } 
        last_digit = digit;
    } while(nr > 0);


    if (streak >= 2) {
       return_value |= STREAK_MIN2;
    }
    if (streak == 2) {
       return_value |= STREAK_EXACT2;
    }

    return return_value;
}

int main()
{
    uint16_t amnt_part1 = 0;
    uint16_t amnt_part2 = 0;
    uint8_t ret;

    for (uint32_t nr = INPUT_MIN; nr <= INPUT_MAX; nr++) {
        ret = match(nr);

        if (ret & STREAK_MIN2) {
            amnt_part1 += 1;
        }
        if (ret & STREAK_EXACT2) {
            amnt_part2 += 1;
        }
    }

    printf("Part 1: %u\n", amnt_part1);
    printf("Part 2: %u\n", amnt_part2);
}
